### DISCUSSION TITLE
Closed Captioning & Accessible SDH Subtitles

### DISCUSSION THREAD

**Discussion by bbgdzxng1** [openai/whisper#458] [Ideas]
I wanted to start a discussion to understand how researchers or app-developers are wrapping Whisper for generating Closed Captioning & SDH Subtitles, since I imagine that accessibility as well as transcription is a common use case.

Whisper generates SRT & WebVTT transcripts by default, producing Pop-on subtitles.

I see that there are some folks pulling word-by-word https://github.com/openai/whisper/discussions/3#discussioncomment-3772777, which I assume is for Paint-on captions.

Can anyone share any lessons-learned or research projects that could be used to generate broadcast-grade captions with Whisper?  When I say broadcast-grade, I mean not just SRT or ASS transcription or translation for home anime consumption, but captions that aim to adhere to the [transcription components of Title 47](https://www.ecfr.gov/current/title-47/chapter-I/subchapter-C/part-79/subpart-A) which describes the US FCC's guidelines for accuracy, synchronicity, completeness and [DCMP-quality captions or subtitles](https://dcmp.org/learn/captioningkey), where the DCMP include recommendations for markup, presentation rate, time-on-screen with a focus for educational use.  The [NIDCD](https://www.nidcd.nih.gov/health/captions-deaf-and-hard-hearing-viewers) also include guidelines for both quality and accessibility.

Obviously, there are limits to how well an AI transcription module can address the accessibility requirements of Closed Captioning (CC) and Subtitles for the Deaf and Hard of Hearing (SDH), but accessibility goes beyond pure transcription.

I see there are active discussions and potential improvements around timing accuracy and timing offset in the repo.

Many of the challenges of the accessibility or standards compatibility elements of Closed Captioning and SDH fall into the domain of a wrapper app, and the requirements get quite complex, quite quickly...
- The common broadcast captioning mechanisms are:
  - Pop-on (line by line, which requires buffering for sentence synchronization)
  - Roll-up
  - Paint-on (word-by-word timecodes)
- I assume that piping/near-realtime of live is unlikely to achieve the level of contextual accuracy and that near-realtime solutions would be limited to adding a tape-delay buffer to allow a look-ahead buffer.  Whisper is more suited to offline, non-realtime use-cases, although some folks have already attempted realtime [here](https://github.com/tobiashuttinger/openai-whisper-realtime) and [here](https://github.com/tekacs/openai-whisper-realtime).
- Closed Captioning Accessibility Best Practice requires a minimum of Speaker Separation by newline (and ideally speaker markup).  This is being addressed in the https://github.com/openai/whisper/discussions/264 thread.  Some broadcasters follow a convention of using `>>` to indicate a change of speaker, whereas the [DCMP encourage parenthesis and discourage characters](https://dcmp.org/learn/603-captioning-key---speaker-identification) (speakername) for identification.
- [Presentation rate](https://dcmp.org/learn/601-captioning-key---presentation-rate), where the words per minute may need to be rate-limited or buffered for readability.  The DCMP Captioning Key recommends a configurable range of {130-160} words per minute.
- Transmission rate, where legacy CC protocols such as SCC & EIA-608 have a hard limit in characters per second/video field
- Character set limitations, where protocols such as EIA-608 are limited to the Title 47 Basic, Special and Extended character sets which may not directly align with ASCII mappings produced by `charset-normalizer`.  A safe filtering option is to use `--suppress_tokens -1`, but broadcast-grade EIA-608 require Extended Characters.
- External libraries for distillation / translation / simplification to "Easy-Reader" Subtitles, where language is simplified to "Third Grade level for language learners".  I'm sure there are some python projects out there that could be used in post process.
- Various timecode formats, which may need to operate in a video frame-based currency, such as `HH:MM:SS;ff` (SMPTE drop frame), `HH:MM:SS:ff` (SMPTE non drop frame) rather than a tick or millisecond currency of `HH:MM:SS.mmm`.  These would need to be handled in any external conversion libraries, such as [pycaption](https://github.com/pbs/pycaption).  Edit: Adding link to discussion topic.  https://github.com/openai/whisper/discussions/1214
- Integration of [profanity detection](https://pypi.org/project/profanity-check/) and [filtering](https://pypi.org/project/profanity-filter/)
- Rule-based linting'n'hinting / validation for the many pro/Broadcast formats such as SCC Sonic Scenarist Captions (the text-based representation of EIA-608), MCC TeleStream MacCaption (the file-based representation of EIA-708), the various XML-based interchange formats (TTML, SMPTE-TT, [EBU-TT](https://ebu.github.io/ebu-tt-live-toolkit/html/validation.html) ) and finally, Whisper's existing webVTT export [should be validated](https://github.com/w3c/webvtt.js) by a wrapper app prior to publishing.  The consumer formats of [SRT](https://github.com/openai/whisper/discussions/199#discussioncomment-3768756) and SSA/ASS have more limited integral accessibility capabilities.
- Proper Noun and dictionary import, without the need for model re-training.
- Language Identification, discussed [here](https://github.com/openai/whisper/discussions/174) and including the language tag in the output filename, `infile.en.vtt`, thereby following the conventions established by mpv and vlc.  It is assumed that Whisper's language detection is superior to [langdetect](https://pypi.org/project/langdetect/).  Language can also be included in a WebVTT header indicating that the timed text track type is captions and the language is English, eg `WEBVTT - This file has cues.; Kind: captions; Language: en`.

All of the above fall outside the direct domain of Whisper, but would fall within the domain of a wrapper service for accessibility for educational broadcast purposes.  Generation of Closed Captioning for educational purposes should aim for the same standards as the regulated broadcast television industry, since one of the tenets of open-education is to increase distribution (through services like over-the-top television) as well as accessibility and inclusiveness.  The open source and FOSS community often enable these solutions without the barrier-to-entry of commercial solutions.

The accuracy of OpenAI's Whisper seems to be some of the best out there, rivaling the current commercial solutions such as [AppTek Appliances](https://www.apptek.com/solutions/live-closed-captioning-appliance), Amazon Rekognition, IBM Watson, Google Speech and Microsoft Azure Speech Service.  Whisper seems far easier to get going compared to Mozilla DeepSpeech, Sphinx or PocketSphinx.  Whisper's transcription accuracy and Proper Noun detection seems to be quite excellent, even with the small models.

Is anyone from research, academic or app-developer communities actively developing Closed Captioning & SDH Subtitle solutions that would improve accessibility for the general public and the community as a whole? 

Thanks in advance for sharing any Whisper projects which fall into the domain of Closed Captioning & Accessibility beyond that of transcription.  And a huge thanks to the developers and sponsors that have contributed to an excellent engine so far.  It will be exciting to watch this project develop and serve the needs of communities who rely on accessibility for educational and entertainment needs.

**Comment by mindofsteel** [comment]
This is an incredible post. I was actually thinking the same thing about a solution that would produce FCC and DCMP-compliant captions.

I love how well you’ve laid this out, and I hope to see some responses from those who might be able to answer your question and/or contribute.

**Comment by Baenwort** [comment]
Closest one so far is https://github.com/openai/whisper/discussions/435 and you might want to reach out to them.

For .srt there has been integration into a software wrapper already but I don't know how much automation they are looking to add: https://github.com/ggerganov/whisper.cpp/discussions/159

**Comment by yqlizeao** [comment]
你的邮件我收到了，稍后给你回复。

**Comment by bbgdzxng1** [comment]
**Re: Subtitle line length.**  Adding a link to a recent improvement from https://github.com/openai/whisper/pull/1184 and the relevant discussion https://github.com/openai/whisper/discussions/314.  The PR implements `--max_line_width` & `--max_line_count`, which will enable a user to limit the line length to 32 character, 4 line ~captions~ subtitles, which helps subsequent conversion from SRT/VTT to SCC or EIA-608.

**Comment by BigShotPunchAndJudy** [comment]
How do I disable the profanity filters tho?

**Comment by skol101** [comment]
Not allowing to disable a profanity filter is a bummer.
